### ISSUE TITLE
Final Issues: Using $.inArray, issues with scope, increasing variables when if returns true

### ISSUE THREAD

**Issue by mlabita37** [mlabita37/Game-Game#2]
I’m pretty close to finished here, just a few kinks needed to work out and I believe I’ll have a really good Battleship game (which I’m calling Ship Strike). As you'll see, I just uploaded the latest version of my game. Here’s the issues I’ve run into that I just can’t seem to figure out on my own:

setComputer ( ):
I’m trying to make sure that the computer does not select duplicate locations. In order to do so, I need to make sure no cell numbers are alike. I’m trying to use jQuery’s $.inArray() function to see whether or not a value is in an array, and if it is, I want to generate a new random location for it. I intend the for loop to run until the location that has been picked doesn’t match any of the prior.  Looking at Line 183 in my code, you’ll see that I split the first computer location then the second one. After that I create a for loop and say: IF either of the numbers from the firstSplit equal any of the numbers in the secondSplit, then generate a new secondSpot. ELSE IF they don’t, then the secondSpot stays the same. The same goes for the thirdSpot below it. 

computerMove ( ) : 
The next issue is with my computerMove function on Line 267. As you see, I first generate a random location on the player’s board to attack. Then I want to compare the location to the player’s location to see whether or not it was a miss. To do so, I split each of the player’s locations, then add them to a new array. I then use this array to compare the computer’s location to, once again using jQuery’s $.inArray() function. Now what I don’t understand is that I applied pretty much the same methodology to the playerMove function above it. The first error I get is that my pValue is undefined. But this is a prototype function, and I’m passing these variables in my declaration. It doesn’t appear to me that there would be any scope issue when using these variables. 

startGame ( ):
One final thing. I’d like to call my setComputer function on Line 180  as part of my start buttons onClick function on Line 214 called startGame().   The only thing is that it’s set to be called when the window loads and I’m declaring a variable there as well. I tried putting this variable at the very end of the startGame function, but I can’t get it to work. Not sure why since I’ve employed the same practice on my early functions.

playGame ( ):
My very last function, it’s a simple function used to return a winner. I’ve chosen to make two variables, playerScore and compScore. They will each start at zero and only increase when a HIT has been logged. As you see, I’ve inserted these into playerMove ( ) on Line 253 and the computerMove on Line 287. My intention is for them to only increase if the IF statement returns true. It doesn’t seem to run at all, though my worry was that it would run every time the for loop runs. Not sure how to approach this one, but the concept is so simple I think I’m just missing one part. NOTE: I have not called this function yet since I cannot get the playerScore or compScore to work yet. I intended on calling it within the startGame ( ) function.

**Comment by clay-reed-a** [comment]
Good Afternoon Michael, 

This is fantastically complicated code. You should be proud of how intricately you are able to think in JavaScript. 

###re: `computerMove ()` 

`pValue` might be `undefined` because 
- I only see `computerMove` called once 
- You are not passing any [arguments](http://stackoverflow.com/questions/1788923/parameter-vs-argument) to when you call it 
- `pValue` is a [parameter](http://stackoverflow.com/questions/1788923/parameter-vs-argument) of `computerMove`

Does that make sense? A prototype function doesn't really behave any differently than a regular function w/ respect to argument variables, and a function definition will overwrite any variables in the parent scope with the same name. 

So:

```
var foo = "foo";

function bar(foo) {
  console.log(foo);
  foo = "bar";
  console.log(foo); 
};

bar(); // prints "undefined" then prints "bar" 
bar(foo); // prints "foo" then prints "bar" 
foo = "bar";
bar(foo); /// prints "bar" and then "bar"  
```

###Now: 

What is the issue w/ `setComputer ( )`? I can't quite tell what you're asking about it.

I also don't understand what is the matter w/ `startGame ( )`? What variable is at issue? `computerLocations`? What's it doing? 

Please advise. Then we can talk about `startGame`. 

Clay a

**Comment by clay-reed-a** [comment]
EDIt: I meant then we can talk about `playGame`.

**Comment by mlabita37** [comment]
Hey Clay, 

Thanks so much for getting back to me! So I'm going through your list trying to weed out the problem in computerMove, here's where I got:
1. So far the value of pValue does not change, so that's not why it's returning undefined later on.
1. I'm only calling computerMove( ) once in the above playerMove( ) function. I just added the parameters: pValue, dValue, sValue to it as well. I tried that at first,  but it the computerMove function still wasn't being called. This time I had to add them as parameters to playerMove() as well. 

3/4. pValue is an argument, but I'm also calling it as a parameter to use it as an argument within other functions. So since it is an argument and part of a prototype function, it should be accessible to all functions. For example:  the pValue is available to the colorPatrolDivs( ) function on Line 133 when called as a parameter, so I'm not sure why the same doesn't apply to the set computer function. 

So here's a repl.it with some changes: 
https://repl.it/B4Xj/0

I added pValue as parameters for computerMove() at the end of the playerMove function(). I also added parameters for playerMove for them to pass. Not sure if that's what you meant. 

Thanks,
Mike

**Comment by clay-reed-a** [comment]
What does pValue stand for?

Clay a

**Comment by mlabita37** [comment]
pValue stands for the value of my first input. In other words, it 's the location of my first ship, the Patrol Boat. That location is stored as a string location, like ("0, 1");  So I need to be able to check this value against the computer's randomly generated move.

**Comment by clay-reed-a** [comment]
Michael, 

I am afraid I am having a little trouble understanding. :-( 

If pValue is always & ever patrolBoat why does it need to be a variable at all? 

Clay a

**Comment by mlabita37** [comment]
I understand your confusion. I hope this helps: The player will choose the Patrol Boat location, or pValue, using a drop-down list. Therefore the pValue will vary depending on what location the player chooses. That is why I need to use pValue later on (I also need use dValue and sValue). I need to know whether the random attack location that the computer chooses is a Patrol Boat, Destroyer, or Submarine location on the player's board. Does that make any sense?

**Comment by mlabita37** [comment]
On another note, I figured out how to get my playerScore variable to increase every time a HIT has been initiated: https://repl.it/B40g

**Comment by mlabita37** [comment]
Also got my setComputer function working properly! It no longer returns duplicates, and if it does, it selects a new random location. Not really sure how I made it work, but I have changed one big thing. I made pValue, dValue, and sValue variables equal to newGame.'function' at the start like so:

window.onload = function() {
  var newGame = new ShipStrikeGame();
  var pValue = newGame.setPatrol();
  var dValue= newGame.setDestroyer();
  var sValue =newGame.setSubmarine();
  var computerLocations = newGame.setComputer();
  newGame.startGame();
  newGame.playerMove(computerLocations);
  var playerScore = newGame.playGame();
  var compScore = newGame.playGame();
};

I'm still not clear as to what this does, so maybe that will help me overcome these last two hurdles?

**Comment by mlabita37** [comment]
Ok, I think I've figured out exactly what's happening. Since I'm calling the playerMove function in my window.onload, it's running before pValue has been defined. I think this is what you meant by what is pValue? Also by your foo/bar example. So what I'm going to attempt to do now is make playerMove( ) run with my startGame function, giving some more meaning to the start game function. I'll also try to get the setComputer function to be called there instead of window.onload, this way the computer board cannot be clicked until the START button has been clicked. Hoping this will do the trick!

**Comment by clay-reed-a** [comment]
Amazing! Let me know when you've pushed to the remote again & I'll take another look. Clay a

**Comment by mlabita37** [comment]
Thanks Clay, I just pushed my most recent version. I think my only issue is now dealing with the scope of my computerLocations variable since it's not being defined in my playerMove( ) function. The computerSet ( ) function, which creates the computerLocations variable, is called before my playerMove( ) function, so I don't understand why it's not being defined. I'm positive it's either a scope issue, or simply an the order in which they're being called but I've tried everything and can't get it to work. Thanks in advance for your help!
